academy/xl-cosmjs/create-custom.md

```diff
@@ -0,0 +1,320 @@
+$ mkdir -p ./proto/gogoproto
+$ curl https://raw.githubusercontent.com/cosmos/cosmos-sdk/v0.42.6/third_party/proto/gogoproto/gogo.proto -o ./proto/gogoproto/gogo.proto
+```
+
+### Compilation
+
+With that, you can now compile the Protobuf files. If you want to avoid adding all the `.proto` files manually to the command, you can use `xargs`, although this is a bit less CPU efficient:
+
+```sh
+$ ls ./proto/myChain | xargs -I {} ./node_modules/protoc/protoc/bin/protoc \
+  --plugin="./node_modules/.bin/protoc-gen-ts_proto" \
+  --ts_proto_out="./client/src/types/generated" \
+  --proto_path="./proto" \
+  --ts_proto_opt="esModuleInterop=true,forceLong=long,useOptionals=messages" \
+  myChain/{}
+```
+
+Notice how `--proto_path` is only `./proto` so that your imports, such as `import "cosmos/base...` can be found.
+
+When it finishes, you should see your files compiled into Typescript. They have been correctly filed under their respective folders and contain both types and services definitions. It also created the compiled versions of your third party imports.
+
+### A note about the result
+
+If your `tx.proto` file contained:
+
+```protobuf [https://github.com/cosmos/cosmos-sdk/blob/v0.42.6/proto/cosmos/bank/v1beta1/tx.proto#L11-L17]
+service Msg {
+      rpc Send(MsgSend) returns (MsgSendResponse);
+      //...
+}
+```
+
+Then, in the compiled `tx.ts` file, you will find its service declaration:
+
+```typescript [https://github.com/confio/cosmjs-types/blob/a14662d/src/cosmos/bank/v1beta1/tx.ts#L243-L248]
+export interface Msg {
+    Send(request: MsgSend): Promise<MsgSendResponse>;
+    //...
+}
+```
+
+And default implementation:
+
+```typescript [https://github.com/confio/cosmjs-types/blob/a14662d/src/cosmos/bank/v1beta1/tx.ts#L250-L268]
+export class MsgClientImpl implements Msg {
+    private readonly rpc: Rpc;
+    constructor(rpc: Rpc) {
+        this.rpc = rpc;
+        this.Send = this.Send.bind(this);
+        //...
+    }
+    Send(request: MsgSend): Promise<MsgSendResponse> {
+        const data = MsgSend.encode(request).finish();
+        const promise = this.rpc.request("cosmos.bank.v1beta1.Msg", "Send", data);
+        return promise.then((data) => MsgSendResponse.decode(new _m0.Reader(data)));
+    }
+    //...
+}
+```
+
+The important points to remember from this are:
+
+1. `rpc: RPC` is an instance of a Protobuf RPC client that will be given to you by CosmJs. Although the interface appears to be [declared locally](https://github.com/confio/cosmjs-types/blob/a14662d/src/cosmos/bank/v1beta1/tx.ts#L270-L272), rthis is the same interface found [througout CosmJs](https://github.com/cosmos/cosmjs/blob/bb92692/packages/stargate/src/queryclient/utils.ts#L35-L37)/ It will be given to you [on construction](https://github.com/cosmos/cosmjs/blob/bb92692/packages/stargate/src/queryclient/queryclient.ts). So, you need not worry about creating an implementation for it.
```

Comment:
```suggestion
1. `rpc: RPC` is an instance of a Protobuf RPC client that will be given to you by CosmJs. Although the interface appears to be [declared locally](https://github.com/confio/cosmjs-types/blob/a14662d/src/cosmos/bank/v1beta1/tx.ts#L270-L272), this is the same interface found [throughout CosmJs](https://github.com/cosmos/cosmjs/blob/bb92692/packages/stargate/src/queryclient/utils.ts#L35-L37). It will be given to you [on construction](https://github.com/cosmos/cosmjs/blob/bb92692/packages/stargate/src/queryclient/queryclient.ts). So, you need not worry about creating an implementation for it.
```